### PR TITLE
Cache exceptions thrown by DiagnosticDescriptor to avoid excessive string allocations

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Removes 27 MB of enumerators:

image

And lowers char arrays from:

image

to

image

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs [hunk 1]
```diff
@@ -17,6 +17,11 @@ internal partial class AnalyzerManager
     {
         private sealed class AnalyzerExecutionContext
         {
+            /// <summary>
+            /// Cached mapping of localizable strings in this this descriptor to any exceptions thrown while obtaining them.
+            /// </summary>
+            private static ImmutableDictionary<LocalizableString, Exception?> s_localizableStringToException = ImmutableDictionary<LocalizableString, Exception?>.Empty.WithComparers(Roslyn.Utilities.ReferenceEqualityComparer.Instance);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs:23]: ❓  Can we use `ImmutableSegmentedDictionary` here? The storage overhead is reduced relative to `ImmutableDictionary`, and the lookup performance (which is the primary fast path) is significantly improved.

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs:23]: Yup yup. I'll make that change.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 2]
```diff
@@ -11792,7 +11792,7 @@ public void LoadinganalyzerNetStandard13()
                 $@"warning AD0001: Analyzer 'TestAnalyzer' threw an exception of type 'System.NotImplementedException' with message '28'.
 System.NotImplementedException: 28
    at TestAnalyzer.get_SupportedDiagnostics()
-   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerManager.AnalyzerExecutionContext.<>c__DisplayClass20_0.<ComputeDiagnosticDescriptors_NoLock>b__0(Object _)
+   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerManager.AnalyzerExecutionContext.<>c__DisplayClass21_0.<ComputeDiagnosticDescriptors_NoLock>b__0(Object _)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:11795]: test depends on exact callstack (including closures) that an exception is reported in.  So a change in the logic there ended up breaking this baseline in a very minor fashion.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler ptal :)

**Review by Youssef1313 — COMMENTED** [review]
Theoretically, different `ToString` calls can produce different exceptions (not common though), but we are caching them now.

Is this a problem?

---

A different approach that seems more straightforward is to have a property `internal Exception? ToStringException { get; private set; }` and `internal bool WasToStringCalled { get; private set; }` in `LocalizableString` and then `AnalyzerManager` would do:

```csharp
if (!localizableString.WasToStringCalled)
{
    _ = localizableString.ToString();
}

if (localizableString.ToStringException is not null)
{
    // ....
}
```

---

I'm quite curious how common exceptions are in `LocalizableString.ToString`?    and what kind of exceptions is this? Can we actually reduce the amount of thrown exceptions in this code path?

**Comment by CyrusNajmabadi** [comment]
> but we are caching them now.

We are only caching the calls to the standard ToString override.

**Comment by CyrusNajmabadi** [comment]
> A different approach 

I preferred not having LocString have any awareness of this. I kept it all at the component that tries to report these exceptions.

**Comment by CyrusNajmabadi** [comment]
> I'm quite curious how common exceptions are in LocalizableString.ToString? 

Very uncommon. 

> and what kind of exceptions is this? 

Can't be known. LocString is a public api. So it's whatever a subclass might decide to throw. 

> Can we actually reduce the amount of thrown exceptions in this code path?

In practice, the number of thrown exceptions is zero. The purpose of this pr though it's to prevent all the extra allocations and overhead we use determining that.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs [hunk 3]
```diff
@@ -17,6 +18,11 @@ internal partial class AnalyzerManager
     {
         private sealed class AnalyzerExecutionContext
         {
+            /// <summary>
+            /// Cached mapping of localizable strings in this this descriptor to any exceptions thrown while obtaining them.
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs:22]:
```suggestion
            /// Cached mapping of localizable strings in this descriptor to any exceptions thrown while obtaining them.
```

**Review by 333fred — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson ptal

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs [hunk 4]
```diff
@@ -17,6 +18,11 @@ internal partial class AnalyzerManager
     {
         private sealed class AnalyzerExecutionContext
         {
+            /// <summary>
+            /// Cached mapping of localizable strings in this descriptor to any exceptions thrown while obtaining them.
+            /// </summary>
+            private static ImmutableSegmentedDictionary<LocalizableString, Exception?> s_localizableStringToException = ImmutableSegmentedDictionary<LocalizableString, Exception?>.Empty.WithComparer(Roslyn.Utilities.ReferenceEqualityComparer.Instance);
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs:24]: There's not a risk of this growing unbounded or anything like that, right?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs:24]: it's effectively capped at "num analyzer * 3".  if we have an unbounded number of analyzers, we're def screwed anyways :)

**Pull request merged**